libraries/AP_Terrain/TerrainIO.cpp

```diff
@@ -151,19 +151,45 @@ void AP_Terrain::open_file(void)
         return;
     }
 
-    // build the pathname to the degree file
-    char path[] = HAL_BOARD_TERRAIN_DIRECTORY "/NxxExxx.DAT";
-    char *p = &path[strlen(HAL_BOARD_TERRAIN_DIRECTORY)+1];
-    snprintf(p, 12, "%c%02u%c%03u.DAT",
-             block.lat_degrees<0?'S':'N',
-             abs(block.lat_degrees),
-             block.lon_degrees<0?'W':'E',
-             abs(block.lon_degrees));
-
-    // create directory if need be
-    if (!directory_created) {
-        mkdir(HAL_BOARD_TERRAIN_DIRECTORY, 0755);
-        directory_created = true;
+    char *path;
+    const char* custom_dir = hal.util->get_custom_terrain_directory();
+    if (custom_dir != NULL){
+        // build the pathname from the args provided
+        path = (char*) malloc(strlen(custom_dir)+strlen("/NxxExxx.DAT")+1);
+        if (!path) {
+            printf("Terrain degree file, malloc() failed: insufficient memory!\n");
+            exit(1);
```

Comment:
You must _never_ call exit() from a function that is called while flying. This will cause the aircraft to fall out of the sky. 
You will need to allocate the path buffer at startup based on length of the custom terrain directory plus room for the degree filename